File: client/androidApp/src/main/java/com/mapmory/android/analytics/FirebaseAnalyticsLogger.kt
package com.mapmory.android.analytics

import android.content.Context
import android.os.Bundle
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.mapmory.shared.analytics.MapmoryAnalytics

class FirebaseAnalyticsLogger(context: Context) : MapmoryAnalytics {
    private val firebaseAnalytics = runCatching {
        FirebaseApp.initializeApp(context.applicationContext)?.let {
            FirebaseAnalytics.getInstance(context.applicationContext)
        }
    }.getOrNull()

    override fun logEvent(name: String, parameters: Map<String, String>) {
        firebaseAnalytics?.logEvent(
            name,
            Bundle().apply {
                parameters.forEach { (key, value) -> putString(key, value) }
            },
        )
    }
}

File: client/iosApp/FirebaseAnalyticsLogger.swift
import FirebaseAnalytics
import Shared

final class FirebaseAnalyticsLogger: NSObject, MapmoryAnalytics {
    func logEvent(name: String, parameters: [String: String]) {
        let firebaseParameters = parameters.reduce(into: [String: Any]()) { result, entry in
            result[entry.key] = entry.value
        }
        Analytics.logEvent(name, parameters: firebaseParameters.isEmpty ? nil : firebaseParameters)
    }
}
